icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,418 @@ namespace icicle {
   public:
     CPUMerkleTreeBackend(
       const std::vector<Hash>& layer_hashes, uint64_t leaf_element_size, uint64_t output_store_min_layer = 0)
-        : MerkleTreeBackend(layer_hashes, leaf_element_size, output_store_min_layer)
+        : MerkleTreeBackend(layer_hashes, leaf_element_size, output_store_min_layer), m_tree_already_built(false)
     {
-      ICICLE_LOG_DEBUG << "in CPUMerkleTreeBackend, have " << layer_hashes.size() << " layer hashes";
-      for (const auto& layer_hash : layer_hashes) {
-        ICICLE_LOG_DEBUG << "layer hash (" << &layer_hash << "), name: " << layer_hash.name();
+      // update layers data base with the hashes
+      const int nof_layers = layer_hashes.size();
+      m_layers.resize(nof_layers);
+      m_pruned_path_size = 0;
+      m_full_path_size = 0;
+      uint64_t nof_hashes = 1;
+      for (int layer_idx = nof_layers - 1; layer_idx >= 0; --layer_idx) {
+        ICICLE_ASSERT(
+          layer_idx == nof_layers - 1 ||
+          layer_hashes[layer_idx + 1].input_default_chunk_size() % layer_hashes[layer_idx].output_size() == 0)
+          << "Each layer output size must divide the above layer input size. Otherwise its not a tree.\n"
+          << "Layer " << layer_idx << " input size = " << layer_hashes[layer_idx + 1].input_default_chunk_size() << "\n"
+          << "Layer " << layer_idx + 1 << " output size = " << layer_hashes[layer_idx].output_size() << "\n";
+
+        // initialize m_layers with hashes
+        auto& cur_layer = m_layers[layer_idx];
+        cur_layer.m_hash = layer_hashes[layer_idx];
+        cur_layer.m_nof_hashes = nof_hashes;
+
+        if (0 < layer_idx) {
+          // update nof_hashes to the next layer (below)
+          const u_int64_t next_layer_total_size = layer_hashes[layer_idx - 1].output_size();
+          nof_hashes = nof_hashes * cur_layer.m_hash.input_default_chunk_size() / next_layer_total_size;
+
+          // Calculate path_size
+          m_pruned_path_size +=
+            layer_hashes[layer_idx].input_default_chunk_size() - layer_hashes[layer_idx - 1].output_size();
+          m_full_path_size += layer_hashes[layer_idx].input_default_chunk_size();
+        }
       }
     }
 
-    eIcicleError build(const std::byte* leaves, uint64_t size, const MerkleTreeConfig& config) override
+    // TODO: handle size
+    eIcicleError build(const std::byte* leaves, uint64_t leaves_size, const MerkleTreeConfig& config) override
     {
-      ICICLE_LOG_DEBUG << "CPU CPUMerkleTreeBackend::build() called with " << size << " bytes of leaves";
-      return eIcicleError::SUCCESS; // TODO: Implement tree-building logic
+      TasksManager<HashTask> task_manager(get_nof_workers(config)); // Run workers.
+      if (m_tree_already_built) {
+        ICICLE_LOG_ERROR << "Tree cannot be built more than one time\n";
```

Comment:
(1) no need to add \n since there is a newline anyway.
(2) maybe it's a warning?

Reply:
\n fixed
its not a warning. I didn't check but it will probably crush if you rebuild